components/raftstore/src/store/fsm/apply.rs

```diff
@@ -3142,333 +2894,311 @@ where
+                        msg
+                    } else {
+                        set_tls_ctx(ctx);
+                        receiver.recv().await
+                    };
+                    if let Some(msg) = msg {
+                        fsm.handle_start = Instant::now_coarse();
+                        ctx = take_tls_ctx();
+                        msg
+                    } else {
+                        return;
+                    }
+                }
+                TryRecvError::Closed => {
+                    return;
+                }
+            },
+        };
+        fsm.handle_task(&mut ctx, msg).await;
     }
 }
 
-impl<EK> DerefMut for ApplyRouter<EK>
+#[derive(Clone)]
+pub struct ApplyRouter<EK>
 where
     EK: KvEngine,
 {
-    fn deref_mut(&mut self) -> &mut BatchRouter<ApplyFsm<EK>, ControlFsm> {
-        &mut self.router
-    }
+    pub inner: FutureSender<Msg<EK>>,
 }
 
 impl<EK> ApplyRouter<EK>
 where
     EK: KvEngine,
 {
-    pub fn schedule_task(&self, region_id: u64, msg: Msg<EK>) {
-        let reg = match self.try_send(region_id, msg) {
-            Either::Left(Ok(())) => return,
-            Either::Left(Err(TrySendError::Disconnected(msg))) | Either::Right(msg) => match msg {
-                Msg::Registration(reg) => reg,
-                Msg::Apply { mut apply, .. } => {
-                    info!(
-                        "target region is not found, drop proposals";
-                        "region_id" => region_id
-                    );
-                    for p in apply.cbs.drain(..) {
-                        let cmd = PendingCmd::<EK::Snapshot>::new(p.index, p.term, p.cb);
-                        notify_region_removed(apply.region_id, apply.peer_id, cmd);
-                    }
-                    return;
-                }
-                Msg::Destroy(_) | Msg::Noop => {
-                    info!(
-                        "target region is not found, drop messages";
-                        "region_id" => region_id
-                    );
-                    return;
-                }
-                Msg::Snapshot(_) => {
-                    warn!(
-                        "region is removed before taking snapshot, are we shutting down?";
-                        "region_id" => region_id
-                    );
-                    return;
-                }
-                Msg::LogsUpToDate(cul) => {
-                    warn!(
-                        "region is removed before merged, are we shutting down?";
-                        "region_id" => region_id,
-                        "merge" => ?cul.merge,
-                    );
-                    return;
-                }
-                Msg::Change {
-                    cmd: ChangeCmd::RegisterObserver { region_id, .. },
-                    cb,
-                    ..
+    pub fn new(sender: FutureSender<Msg<EK>>) -> ApplyRouter<EK> {
+        ApplyRouter { inner: sender }
+    }
+    pub fn schedule(&mut self, msg: Msg<EK>) {
```

Comment:
Why changing the original `schedule_task` implementation?

Reply:
Because we can send a msg to `receiver` of a region directly. The method `try_send` will notify the future so we do not need `batch-system` and `Applyrouter`.

Reply:
There are some callbacks in the `ApplyTask`, I think we should `notify_region_removed` for them in the case of `TrySendError::Closed`.

Reply:
I have implement `Drop` for `ApplyFsm`, and all of `callback` will be 'notify_stale_command' when the fsm is dropped.

Reply:
Oh, sorry. I misunderstand what you point. I will fix it.

Reply:
In the previous handling, `Msg::Change` also has a callback needs to invoke. Maybe we can just keep the previous handling as there are some logs maybe be helpful when debugging.